Unique stream consumer name per service and stage

preshetin opened the pull request:
Closes: #9491

## How to test it

1. Use **serverless.yaml** with Kinesis steam that has `consumer` parameter.

<details>
<summary>Example serverless.yaml</summary>

```
service: sls-kinesis
frameworkVersion: '2'

provider:
  name: aws
  runtime: nodejs12.x
  lambdaHashingVersion: 20201221
  kinesis:
    consumerNamingMode: 'serviceSpecific'

functions:
  hello:
    handler: handler.hello
    events:
      - stream:
          type: kinesis
          
          # Create your own Kinesis stream by running `aws kinesis create-stream --stream-name Foo --shard-count 1 `. 
          # Don't forget to remove it after testing is finished with `aws kinesis delete-stream --stream-name Foo`
          arn: arn:aws:kinesis:us-east-1:0123456789:stream/Foo
          consumer: true
```
</details>

3. Deploy it two times, on different stagings `dev` and `another` with the following commands
-  `sls deploy -v --stage dev` 
-  `sls deploy -v --stage another`
4. Before applying this fix, the second deployment would result in an error like this:

```
 Serverless Error ----------------------------------------
 
  An error occurred: HelloFooConsumerStreamConsumer - helloFooConsumer|arn:aws:kinesis:us-east-1:0123456789:stream/Foo already exists in stack arn:aws:cloudformation:us-east-1:0123456789:stack/sls-kinesis-dev/3a3471a0-df08-11eb-ba9a-0afdb21fa153.
```

5. After applying this fix, the second deployment should succeed.

preshetin commented:
Hey @aheuermann it would be great if you have a chance to test it. This PR fixes the issue you opened some time ago

pgrzesik reviewed (commented):
thanks for working on this PR @preshetin :bow: I'm going to hold off with review until we got everything sorted out in the issue's discussion

preshetin commented:
> I'm going to hold off with review until we got everything sorted out in the issue's discussion

I think we sorted out the implementation so I'm hoping @pgrzesik it is ready for review.

One thing I noticed in failing checks is some tests are missing. I can add them. Just want to make sure the rest is fine.

Also, I'm adding deprectations for the first time so I'm not sure I implemented it 100% property. The message does show in console output when the new naming property is set but I think additional docs have to be added at `/docs/deprecations.md` page. I'll add them.

pgrzesik commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -374,6 +374,11 @@ module.exports = {
   getStreamConsumerName(functionName, streamName) {
     return `${functionName}${streamName}Consumer`;
   },
+  getServiceSpecificStreamConsumerName(functionName, streamName) {
```
it would be good to cover this with unit test

pgrzesik replied:
Additionally, I think it would be better to not have separate util for that but go with the similar route to `getHttpApiName` for example with encapsulating that logic inside

preshetin replied:
Yes, I'll add tests and will try to make it similar to `getHttpApiName `

pgrzesik commented on lib/plugins/aws/package/compile/events/stream.js:
```diff
@@ -273,10 +286,21 @@ class AwsCompileStreamEvents {
+                if (
+                  providerKinesisConfig &&
+                  providerKinesisConfig.consumerNamingMode &&
+                  providerKinesisConfig.consumerNamingMode === 'serviceSpecific'
```
It would be good to cover these two cases with separate tests

pgrzesik replied:
We should also add information about that property to documentation about Kinesis

preshetin replied:
> It would be good to cover these two cases with separate tests

@pgrzesik can you give a hint on how to create a test? I'm trying to make use of `runServerless` command when adding a test like this:

```javascript

describe('Naming Mode Deprecation', () => {
  it('should show deprecation message when old naming is used', async () => {
    await runServerless({
      fixture: 'function',
      configExt: {
        functions: {
          foo: {
            events: [{
              stream: {
                type: 'kinesis',
                arn: 'arn:aws:kinesis:us-east-1:027546143534:stream/Foo',
                consumer: true
              },
            }],
          },
        },
      },
      command: 'package',
    });
    // TODO: assert that stdoutData contains deprecation message
  });
});
```

when I run this test it fails with this error: 

```
% npx mocha /Users/petrreshetin/Code/serverless-serverless/test/unit/lib/plugins/aws/package/compile/events/stream.test.js --grep Deprecation                     


  Naming Mode Deprecation
    1) should show deprecation message when old naming is used
2021-09-06T13:21:10.722Z ✖ mocha test fail ServerlessError: consumerNamingMode will be different in v3. The new naming will cause redeployment which may lead to delay in data processing\n' +
  '  More Info: https://www.serverless.com/framework/docs/deprecations/#consumerNamingMode\n' +
  '    at module.exports (lib/utils/logDeprecation.js:88:15)\n' +
```

I tried to wrap this test with expect(...).to.throw(ServerlessError) however it does not work

pgrzesik replied:
You can ignore deprecations with `runServerless` runs with `disabledDeprecations` passed into `configExt` so it's still possible to test old logic. As for testing if deprecation is displayed, I think it can be skipped

preshetin replied:
> It would be good to cover these two cases with separate tests

Since there is already a test conving standard naming case I just added another one handling the new naming mode case

pgrzesik commented on lib/plugins/aws/package/compile/events/stream.js:
```diff
@@ -9,6 +9,19 @@ class AwsCompileStreamEvents {
 
     this.hooks = {
+      'initialize': () => {
+        if (
```
I think this deprecation does not address the issue that we would like to address with the message. We don't really care about people who have this setting turned on, but rather about people that will be affected by the change with the next major as it might be a breaking change, so instead, we should check for people that have `consumer` set but do not use `kinesis.consumerNamingMode` that in v3 the naming scheme will change and they can already opt in by setting `kinesis.consumerNamingMode`.

preshetin replied:
I agree, I'll have to redo this

pgrzesik commented on lib/plugins/aws/package/compile/events/stream.js:
```diff
@@ -9,6 +9,19 @@ class AwsCompileStreamEvents {
+          this.serverless.service.provider.kinesis.consumerNamingMode &&
+          this.serverless.service.provider.kinesis.consumerNamingMode === 'serviceSpecific'
+        ) {
+          this.serverless._logDeprecation(
```
We should document this in `deprecations.md` doc - please see the short guide about configuring deprecations: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#configuring-deprecations

preshetin replied:
Will do this

pgrzesik reviewed (changes requested):
Thanks @preshetin - overall it looks good, I have a few comments though, please let me know what do you think :bow:

preshetin commented:
Thanks, @pgrzesik for the review 🙇 . I tend to agree with all your points. I'll address them and will get back to you

preshetin commented:
Hi, @pgrzesik - I've addressed your comments from the last review. 

Let me know what you think

preshetin commented:
Hey @pgrzesik - no rush with that. I'm also planning to be away in the next few weeks

pgrzesik commented on docs/deprecations.md:
```diff
@@ -354,6 +354,18 @@ Adapt to this convention now by setting `provider.apiGateway.shouldStartNameWith
 
+<a name="KINESIS_CONSUMER_NAME_CONTAINING_SERVICE"><div>&nbsp;</div></a>
+
+## Kinesis consumer name will be changed which will lead downtime during re-deployment
```
Instead of writing of what might happen in the title, let's just stick to something like 

```
## Kinesis consumer naming will be changed to ...
```

and in description mention what might happen during the change (e.g. short downtime), which you already added in the description and it sounds good in my opinion :+1:

pgrzesik commented on lib/plugins/aws/package/compile/events/stream.js:
```diff
@@ -9,6 +9,26 @@ class AwsCompileStreamEvents {
+        ) {
+          this.serverless._logDeprecation(
+            'KINESIS_CONSUMER_NAME_CONTAINING_SERVICE',
+            'Consumer name for Kinesis streams will be different in v3. The new naming will cause re-deployment which may lead to delay in Kinesis stream data processing.'
```
Let's be a little bit more clear here and follow similar messaging to already existing deprecation notices. 

Maybe something along those lines? (It's definitely not a final version) 

```
Starting with v3.0.0, generated consumer name for Kinesis streams will use a new naming scheme that will include service and stage to ensure uniqueness across different services and/or stages. In order to hide the deprecation and ensure seamless upgrade, please set the property ...
```

pgrzesik reviewed (changes requested):
Thanks @preshetin - looks good in general, I have a few comments related to reporting the deprecation and we should be good to go

preshetin commented:
Thanks @pgrzesik for the review. I'm still on volunteering vacation. Hoping to take a look at this PR by the end of October

preshetin commented:
Hey @medikoo - I updated this PR with the latest v3 code. I also tested it locally and can confirm it fixes the issue.  

I also ran tests locally and can confirm they pass except the one that [can by ignored](https://github.com/serverless/serverless/pull/9706#discussion_r702944028). Some tests are pending which I believe are related to Python runtime. 

I'm asking for some guidance on the deprecation message approach. The deprecation message currently contains `Starting with v3.0.0` which I believe should be updated to something more relevant.

I did not use Serverless Framework for some time. If I remember last time I used it it was v2. (Personally switched to SST.) So I may miss something.

Looking for your feedback

medikoo reviewed (changes requested):
Great thanks @preshetin for the update!

I've posted my comments, in general it looks very good.

preshetin commented:
Thanks for the feedback @medikoo 

I've addressed all your comments. Now all tests pass.

Looks like this is ready for merging. Otherwise feel free to share your concerns

medikoo commented on test/unit/lib/plugins/aws/package/compile/events/stream.test.js:
```diff
@@ -1479,6 +1479,11 @@ describe('test/unit/lib/plugins/aws/package/compile/events/stream.test.js', () =
         configExt: {
+          provider: {
+            kinesis: {
+              consumerNamingMode: 'serviceSpecific',
```
Can we confirm that new naming has picked up in one of the underlying tests?

preshetin replied:
Yes, I can confirm that. Without these lines one test is failing:

```
% npx mocha test/unit/lib/plugins/aws/package/compile/events/stream.test.js

....

  24 passing (3s)
  2 pending
  1 failing

  1) test/unit/lib/plugins/aws/package/compile/events/stream.test.js
       regular
         "before all" hook for "TODO: should support ARN String for `arn`":
     ServerlessError: Starting with v4.0.0 Kinesis streams naming scheme will change. Adapt to new schema by setting`provider.kinesis.consumerNamingMode` property to `serviceSpecific`.
  More Info: https://www.serverless.com/framework/docs/deprecations/#KINESIS_CONSUMER_NAME_CONTAINING_SERVICE
```

After adding them, all tests pass.

medikoo replied:
@preshetin I suggested adding below a test, in which we confirm that, indeed new naming is used.

Currently, we've introduced a feature that doesn't have a single confirmation test, and that's not great

preshetin replied:
@medikoo  I think I get what you mean. Sure, I'll do this and let you know.

preshetin replied:
@medikoo test added in https://github.com/serverless/serverless/pull/9706/commits/fa7282a102c76a468a59552d96fa839fe9540d85 - should it be like that?

medikoo replied:
@preshetin We should rather access the generated CF resource, and confirm that consumer name has the _service_ name in it.

Additionally _service_ name in case of `runServerless` is generated automatically, so instead of enforcing its value via `service` property, it'll be nicer to simply retrieve it from returned `serverless` instance (you can see how it's accessed here: https://github.com/serverless/serverless/blob/34d922d3d8d33f6e17e697610fdef7f13003a2f3/test/unit/lib/plugins/aws/remove/index.test.js#L218-L241)

preshetin replied:
Thanks for the hints @medikoo. Here is another attempt at https://github.com/serverless/serverless/pull/9706/commits/6184922931ca7361991574c6a5247267b50230a9

medikoo reviewed (commented):
Thank you @preshetin !

medikoo reviewed (approved):
That looks great @preshetin, it looks really good now. Great thanks for being willing to finalize this. It's highly appreciated 🙇
